src/System.Security.Cryptography.X509Certificates/src/System/Security/Cryptography/X509Certificates/CertificateRequest.cs

```diff
@@ -365,6 +365,85 @@ public X509Certificate2 CreateSelfSigned(DateTimeOffset notBefore, DateTimeOffse
+            long notAfterLocalTicks = notAfterLocal.Ticks;
+            long fractionalSeconds = notAfterLocalTicks % TimeSpan.TicksPerSecond;
+            notAfterLocalTicks -= fractionalSeconds;
+            notAfterLocal = new DateTime(notAfterLocalTicks, notAfterLocal.Kind);
+
+            if (notAfterLocal > issuerCertificate.NotAfter)
+            {
+                throw new ArgumentException(
+                    SR.Format(
+                        SR.Cryptography_CertReq_NotAfterNotNested,
+                        notAfterLocal,
+                        issuerCertificate.NotAfter),
+                    nameof(notAfter));
+            }
+
+            // Check the Basic Constraints and Key Usage extensions to help identify inappropriate certificates.
+            // Note that this is not a security check. The system library backing X509Chain will use these same criteria
+            // to determine if a chain is valid; and a user can easily call the X509SignatureGenerator overload to
+            // bypass this validation.  We're simply helping them at signing time understand that they've
+            // chosen the wrong cert.
+            X509BasicConstraintsExtension basicConstraints = null;
+            X509KeyUsageExtension keyUsage = null;
+            X509SubjectKeyIdentifierExtension skid = null;
+            X509AuthorityKeyIdentifierExtension akid = null;
+
+            foreach (X509Extension extension in issuerCertificate.Extensions)
+            {
+                if (extension.Oid.Value == Oids.BasicConstraints2)
+                {
+                    basicConstraints = (X509BasicConstraintsExtension)extension;
+                }
+                else if (extension.Oid.Value == Oids.KeyUsage)
+                {
+                    keyUsage = (X509KeyUsageExtension)extension;
+                }
+                else if (extension.Oid.Value == Oids.SubjectKeyIdentifier)
+                {
+                    skid = (X509SubjectKeyIdentifierExtension)extension;
+                }
+                else if (extension.Oid.Value == Oids.AuthorityKeyIdentifier)
+                {
+                    akid = new X509AuthorityKeyIdentifierExtension();
+                    akid.CopyFrom(extension);
+                }
+
+                // Note that skid and akid are not present here.
+                // If we find the basic constraints extension we don't need the other two.
+                if (keyUsage != null && basicConstraints != null)
+                {
+                    break;
+                }
+            }
+
+            if (basicConstraints != null && !basicConstraints.CertificateAuthority)
+                throw new ArgumentException(SR.Cryptography_CertReq_IssuerBasicConstraintsInvalid, nameof(issuerCertificate));
+
+            // Windows only allows Basic Constraints to be missing for a self-signed cert.
+            if (basicConstraints == null && !issuerCertificate.IsSelfSigned(skid, akid))
```

Comment:
Assert or check for null `skid`, `akid` here? Or just let NullReferenceException?